cirq/neutral_atoms/neutral_atom_devices.py

```diff
@@ -0,0 +1,424 @@
+
+    def validate_moment(self, moment: ops.Moment):
+        """
+        Raises an error if the given moment is invalid on this device
+
+        Args:
+            moment: The moment to validate
+
+        Raises:
+            ValueError: If the given moment is invalid
+        """
+        super().validate_moment(moment)
+
+        operations = moment.operations
+        num_parallel_z = 0
+        num_parallel_xy = 0
+        num_parallel_controlled_qubits = 0
+        hasMeasurement = False
+        controlled_qubits_lists = []
+
+        # Count the number of each type of gate occuring in this moment.
+        # Also verify that gates of the same type are equal
+        for op in operations:
+            if isinstance(op, (ops.GateOperation,
+                               ops.ParallelGateOperation)):
+                if isinstance(op.gate, ops.ZPowGate):
+                    if num_parallel_z == 0:
+                        refZgate = op.gate
+                    else:
+                        if refZgate != op.gate:
+                            raise ValueError("Non-identical Parallel Z gates")
+                    num_parallel_z += len(op.qubits)
+
+                if isinstance(op.gate, (ops.XPowGate,
+                                        ops.YPowGate,
+                                        ops.PhasedXPowGate)):
+                    if num_parallel_xy == 0:
+                        refXYgate = op.gate
+                    else:
+                        if refXYgate != op.gate:
+                            raise ValueError("Non-identical Parallel XY gates")
+                    num_parallel_xy += len(op.qubits)
+
+                if isinstance(op.gate, (ops.CNotPowGate,
+                                        ops.CZPowGate,
+                                        ops.CCXPowGate,
+                                        ops.CCZPowGate)):
+                    if num_parallel_controlled_qubits == 0:
+                        refCgate = op.gate
+                    else:
+                        if refCgate != op.gate:
+                            raise ValueError("Non-identical Parallel Controlled"
+                                             " Gates")
+                    num_parallel_controlled_qubits += len(op.qubits)
+                    controlled_qubits_lists.append(op.qubits)
+                if isinstance(op.gate, ops.MeasurementGate):
+                    hasMeasurement = True
+
+        if num_parallel_controlled_qubits > 0:
+            if num_parallel_controlled_qubits > self._max_parallel_c:
+                raise ValueError("Too many qubits acted on by controlled gates")
+            if num_parallel_xy > 0 or num_parallel_z > 0:
+                raise ValueError("Can't perform non-controlled operations at"
+                                 " same time as controlled operations")
+            for op1num, op1qubits in enumerate(controlled_qubits_lists):
```

Comment:
Extract this into a private instance or module level method `are_qubit_lists_too_close` that takes a varargs `*qubit_lists`. If the number of them is 0 or 1, return false. If it's more than 2, iterate over all the pairs recursing on the same method. if it's 2, go over the qubit pairs.

Reply:
Done